edi/models/edi_exchange_consumer_mixin.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2020 ACSONE SA
+# Copyright 2020 Creu Blanca
+# @author Simone Orsi <simahawk@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+
+
+class EDIExchangeConsumerMixin(models.AbstractModel):
+    """Record that might have related EDI Exchange records
+    """
+
+    _name = "edi.exchange.consumer.mixin"
+    _description = "Abstract record where exchange records can be assigned"
+
+    exchange_record_ids = fields.One2many(
+        "edi.exchange.record",
+        inverse_name="res_id",
+        domain=lambda r: [("model", "=", r._name)],
+    )
+    exchange_record_count = fields.Integer(compute="_compute_exchange_record_count")
+    missing_exchange_records = fields.Boolean(
+        compute="_compute_has_missing_exchange_records"
+    )
+
+    def _get_exchange_record_configuration(self):
+        """ This function should return a dict containing backend and exchange_type.
+        Example::
+        {
+            'configuration_identifier': {
+                'exchange_type': 'exchange_type_code', # required, string
+                'backend': backend # required, a `edi.backend` object,
+                'create_vals': {} # not required, dict with extra values for the
+                                  # creation vals
+            }
+        }
+        To be overridden by the modules.
+        We might need to check if the record has already been generated.
+        You can check an example on the test
+        """
+        self.ensure_one()
+        return {}
+
+    def _has_exchange_record(self, exchange_type, backend, extra_domain=False):
+        """This function is useful when generating the configuration"""
+        domain = [
+            ("model", "=", self._name),
+            ("res_id", "=", self.id),
+            ("backend_id", "=", backend.id),
+            ("type_id.code", "=", exchange_type),
+        ]
+        if extra_domain:
+            domain += extra_domain
+        return self.env["edi.exchange.record"].search(domain, limit=1)
+
+    @api.depends("exchange_record_ids")
+    def _compute_has_missing_exchange_records(self):
+        for record in self:
+            record.missing_exchange_records = bool(
+                record._get_exchange_record_configuration()
+            )
+
+    def _generate_exchange_records(self):
+        """This function will generate the exchange records of consumers.
+
+        For example, this can be added when posting an invoice or validating a picking
+        """
+        exchange_records = self.env["edi.exchange.record"]
```

Comment:
@etobella as we discussed, let's leave here only the hook `_generate_exchange_records`.

Reply:
I think, maintaining it has not much sense, I will keep the `_has_exchange_record`

Reply:
It has sense as we still want to support a button to generate it manually.
I'd have `edi_generate_records` (public, can be used by a btn)  and `_edi_generate_records` which by default here raises a `NotImplementedError`.

Reply:
Done